### PR TITLE
Certify Azure Event Hubs as Stable in pubsub

### PR DESCRIPTION
**Pull request by surenderssm**
Signed-off-by: Surender Singh Malik <surenderssm@users.noreply.github.com>

# Description

_Certifications test for azure eventhubs to promote the same as stable pubsub component_

## Pending Items (work in progress)

- Test with IOT hub is pending for integration as here we need to use `az cli` to emulate a IOT device sending messages. ( subscribe part tested locally, code is commented)

## Future Work
- Test for network interruptions while publishing messages [known issue with eventhubs component]
## Issue reference

Please reference the issue this PR will close: #1379

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

## Testing

- Test cases were run and verified in local by updating the metadata with actual azure resources like SPN, Storage account, eventhubs, iot hub.

### CODE REVIEW

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+				s.AddTopicEventHandler(&common.Subscription{
+					PubsubName: pubsubName,
+					Topic:      topicName,
+					Route:      "/orders",
+				}, func(_ context.Context, e *common.TopicEvent) (retry bool, err error) {
+					if err := sim(); err != nil {
+						return true, err
+					}
+
+					fmt.Println(e.Data)
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:103]: can be removed ...

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort+portOffset*3),
+			embedded.WithDaprGRPCPort(runtime.DefaultDaprAPIGRPCPort+portOffset*3),
+			embedded.WithDaprHTTPPort(runtime.DefaultDaprHTTPPort+portOffset*3),
+			embedded.WithProfilePort(runtime.DefaultProfilePort+portOffset*3),
+			runtime.WithPubSubs(component))).
+		Step("publish messages to topic1", publishMessages(metadata, sidecarName4, topicToBeCreated, consumerGroup4)).
+		Step("Verify if app5 has recevied messages published to topic", assertMessages(50*time.Second, consumerGroup4)).
+
+		// IOT Test : app 5
+		// PREREQ : Add messages to IOT endpoint via (./send-iot-device-events.sh)
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:233]: is it a pre req that `./send-io-device-events.sh` is run before this test ? 
Can that logic be pulled into this test and be setup in a step here?

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort+portOffset*2),
+			embedded.WithDaprGRPCPort(runtime.DefaultDaprAPIGRPCPort+portOffset*2),
+			embedded.WithDaprHTTPPort(runtime.DefaultDaprHTTPPort+portOffset*2),
+			embedded.WithProfilePort(runtime.DefaultProfilePort+portOffset*2),
+			runtime.WithPubSubs(component))).
+
+		// publish message again to topic1, however this time there are more than one consumer group on same topic
+		Step("publish messages to topic1", publishMessages(metadata, sidecarName1, topicName1, consumerGroup2)).
+		Step("publish messages to topic1", publishMessages(metadata, sidecarName2, topicName1, consumerGroup2)).
+		Step("Verify if app1, app3 together has recevied messages published to topic1", assertMessages(10*time.Second, consumerGroup2)).
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:215]: quick question ... if only a single partition is used, how can two consumers consume from the same partition here? 
```
metadata := map[string]string{
 		messageKey: partition0,
 	}
```

Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:215]: nit isn't this `app1 and app2` ? instead of app3?

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 2 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:233]: Is it a pre req that ./send-io-device-events.sh is run before this test ?
- Yes

Can that logic be pulled into this test and be setup in a step here?

- Yes that's one option, second option is to pull this as part of `github workflow`. Option one sounds closer to test, will give it a shot.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 3 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:215]: Good catch, currently `package watcher` is agnostic of number of owners, hence the test passes.

It's a miss, will publish for one more set of partition key.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 1 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: Fixed.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 3 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: Fixed.
Test : multiple publisher with different partition key, multiple subscriber with same consumer ID
app2, app3 are two subscribers to single topic with same consumer ID.

Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: Fixed.

**Comment by surenderssm** [comment]
@daixiang0 - Thank you for your note, PR is ready for review (removed WIP tag). Do let me know your valuable inputs on this change. 
> @surenderssm thanks for your contribution! Please remove WIP tag when ready for review.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort+portOffset),
+			embedded.WithDaprGRPCPort(runtime.DefaultDaprAPIGRPCPort+portOffset),
+			embedded.WithDaprHTTPPort(runtime.DefaultDaprHTTPPort+portOffset),
+			embedded.WithProfilePort(runtime.DefaultProfilePort+portOffset),
+			runtime.WithSecretStores(secretStoreComponent),
+			runtime.WithPubSubs(component))).
+		Step("publish messages to topic1", publishMessages(metadata, sidecarName1, topicName1, consumerGroup1, consumerGroup2)).
+		// TODO : option 1: Give sidecar1 component refering eh connection string with acess on both the topics
+		// option 2: Start a sidecar with a component refering eh connection string of unusedTopic
+		// Step("publish messages to unUsedTopic", publishMessages(metadata, sidecarName1, unUsedTopic)).
```
Comment (surenderssm) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:209]: This is in my radar will be picking this up in next commit

Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: Fixed.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+		Step(sidecar.Run(sidecarName2,
+			embedded.WithComponentsPath("./components/consumer2"),
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort+portOffset),
+			embedded.WithDaprGRPCPort(runtime.DefaultDaprAPIGRPCPort+portOffset),
+			embedded.WithDaprHTTPPort(runtime.DefaultDaprHTTPPort+portOffset),
+			embedded.WithProfilePort(runtime.DefaultProfilePort+portOffset),
+			runtime.WithSecretStores(secretStoreComponent),
+			runtime.WithPubSubs(component))).
+
+		// Run dummy app
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:228]: A dummy app is not needed ideally to publish to another topic. 
It does not matter to the publisher whether the message is consumed or not ... 
You can make use of the same sidecar sidecar1 itself to publish to another topic which has no subscribers ... Why is this needed?

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+			embedded.WithComponentsPath("./components/dummyconsumer"),
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort+portOffset*10),
+			embedded.WithDaprGRPCPort(runtime.DefaultDaprAPIGRPCPort+portOffset*10),
+			embedded.WithDaprHTTPPort(runtime.DefaultDaprHTTPPort+portOffset*10),
+			embedded.WithProfilePort(runtime.DefaultProfilePort+portOffset*10),
+			runtime.WithSecretStores(secretStoreComponent),
+			runtime.WithPubSubs(component))).
+		Step("publish messages to topic1", publishMessages(nil, sidecarName1, topicName1, consumerGroup1, consumerGroup2)).
+		// Start a sidecar with a component refering eh connection string of unusedTopic
+		Step("publish messages to unUsedTopic", publishMessages(nil, sidecarDummyName, unUsedTopic)).
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:242]: you can use sideCar1 itself to publish to a different topic ... ? Why is the dummy sidecar needed?

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+		Step("verify if app1 has recevied messages published to topic1", assertMessages(10*time.Second, consumerGroup1)).
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:243]: is negative assertion possible ? 
To verify it has not received any messages from the unused topic?

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+		// publish message in topic1 from two publisher apps, however there are two subscriber apps (app2,app3) with same consumerID
+		Step("publish messages to topic1", publishMessages(metadata, sidecarName1, topicName1, consumerGroup2)).
+		Step("publish messages to topic1", publishMessages(metadata1, sidecarName2, topicName1, consumerGroup2)).
+		Step("verify if app2, app3 together have recevied messages published to topic1", assertMessages(10*time.Second, consumerGroup2)).
+
+		// Test : Entitymanagement
+		// Run subscriberApplication app4
+		Step(app.Run(appID4, fmt.Sprintf(":%d", appPort+portOffset*3),
+			subscriberApplication(appID4, topicToBeCreated, consumerGroup4))).
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:270]: when is this topic deleted? if this test needs to test entity creation every time. otherwise it will not be creating that event hub but reuse the existing event hub created when the first test was run ...

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+   - App: Simulates periodic errors
+   - Component: Retries on error
+   - App: Observes successful messages
+   - Test: Confirms that all expected messages were received
+   - Test: Confirms that all expected system properties are set
+### Authentication Tests 
+  - Test connection string based authentication mechanism
+  - Test AAD Service Principal based authentication
+    - Utilize a service principal with appropriate roles granted
+  - Test MSI based authentication
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/README.md:69]: mark MSI as TODO

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    - Utilize Managed Identity
+### Network Tests?
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/README.md:71]: mark network tests as TODO

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    - App: Observes successful messages
+    - Test: Confirms that all expected messages were received
+    - Test: Confirms that each subscriber receives messages from both partitions
+ - Test entity management
+   - Start a publisher and subscriber with a topic that does not exist
+   - Test: Confirm creation of topic/eventHub in given eventHub namespace
+   - Test: Send 100+ unique messasges to the newly created eventHub topic
+   - App: Observe successful messages
+   - Test: Confirm that subscriber receives all messages
+ - Test IOT Event Hub
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/README.md:55]: mark as TODO

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
+# Azure Event Hubs Pubsub certification testing
+
+This project aims to test the Azure Event Hubs pubsub component under various conditions.
+
+## Test Plan
+
+### Pubsub Certification Tests
+  - Test partition key, in order processing single publisher/subscriber (done as part of conformance tests)
+  - Test single publisher / single subscriber
```
Comment (mukundansundar) [inline tests/certification/pubsub/azure/eventhubs/README.md:9]: This test is not created yet.

**Comment by mukundansundar** [comment]
@surenderssm Any updates on the PR?

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 8 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:272]: Thanks for the catch, this has been fixed as part of cleanup step up in the test.

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 9 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md]: Done.

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 10 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md:71]: Done.

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 12 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md:9]: This is being tested with along with entity management test.
- create topic-> publish messages-> subscribe messages-> check if the messages were received in order.

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    - Start an app with 1 publisher and 1 subscriber
+    - The publisher publishes to 2 topics 
+    - The subscriber is subscribed to 1 topic
+    - Test: Sends 100+ unique messages
+    - App: Simulates periodic errors
+    - Component: Retries on error
+    - App: Observes successful messages
+    - Test: Confirms that all expected messages were received
+    - Test: Confirms that subscriber does not receive messages from the non-subscribed topic
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md:18]: Currently the platform package `watcher` do not have the capabilities to enable this scenario. 
We can take a look at this in next version of the PR. 

Any thoughts?

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 11 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md]: done.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 6 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:244]: eh connectionstring provided in component (`consumer1\eventhubs.yaml`) which is loaded by sideCar1 is scoped at topic level. With this constraint it won't be possible to use sidecar1 to publish to a different topic.

For reference : (EntityPath part of connection string : topic name)

`"Endpoint=sb://*.servicebus.windows.net/;SharedAccessKeyName=**;SharedAccessKey=*;EntityPath=certification-pubsub-topic1"`

This will be possible only if provide connection string from the policy defined at namespace level.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 7 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:245]: Package `watcher` being used here (also in kafka) ignores the messages which are not part of expected strings. 

One possible solution i can think of is exposing actual & expected arrays data types via changing the scope or a different function.

Once we have access we can run set operations to achieve our goal.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 6 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:244]: Notes from offline conversation with Mukundan.

- Test for connection string at namespace level.
- Publishing to multiple topics with a single sidecar.
	- subscription to multiple topics with single sidecar
	- Sidecar 1 : namespace level
		- Multiple topics

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 12 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md:9]: Added description in form of comment.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 6 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: Fixed.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 5 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: Fixed.

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (mukundansundar) [reply tests/certification/pubsub/azure/eventhubs/README.md:18]: as discussed please create a separate issue for enhancement ...

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 7 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:222]: To be tracked as part of : https://github.com/dapr/components-contrib/issues/1663

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 2 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:264]: to be tracked as part of : https://github.com/dapr/components-contrib/issues/1664

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+		// Run the Dapr sidecar with the component entitymanagement
+		Step(sidecar.Run(sidecarName4,
+			embedded.WithComponentsPath("./components/entitymanagementconsumer"),
+			embedded.WithAppProtocol(runtime.HTTPProtocol, appPort+portOffset*3),
+			embedded.WithDaprGRPCPort(runtime.DefaultDaprAPIGRPCPort+portOffset*3),
+			embedded.WithDaprHTTPPort(runtime.DefaultDaprHTTPPort+portOffset*3),
+			embedded.WithProfilePort(runtime.DefaultProfilePort+portOffset*3),
+			runtime.WithSecretStores(secretStoreComponent),
+			runtime.WithPubSubs(component))).
+		Step("publish messages to topicToBeCreated", publishMessages(metadata, sidecarName4, topicToBeCreated, consumerGroup4)).
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go]: fixed.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+
+	deleteEventhub := func(ctx flow.Context) error {
+		output, err := exec.Command("/bin/sh", "delete-eventhub.sh", topicToBeCreated).Output()
+		assert.Nil(t, err, "Error in delete-eventhub.sh.:\n%s", string(output))
+		return nil
+	}
+
+	flow.New(t, "cleanup azure artifacts").
+		Step("delete eventhub created as part of the eventhub management test", deleteEventhub).
+		Run()
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:287]: yes, any failures above will prevent this from running.

One option was to run this before the step where eh is getting created, to ensure a clean state.
However in my test in local, I observed that probability of failures increases (timeout) on calling delete, followed by create one after the other.

Any suggestions how to proceed on this?

**Review by mukundansundar — DISMISSED** [review]
lgtm

**Comment by mukundansundar** [comment]
@berndverst Please review this PR ...

File: .github/infrastructure/conformance/azure/conf-test-azure-eventhubs.bicep [hunk 16]
```diff
@@ -61,6 +85,39 @@ resource eventHubsNamespace 'Microsoft.EventHub/namespaces@2017-04-01' = {
...
+      }
+    }
+    resource eventHubPubsubConsumerGroup1 'consumergroups' = {
+      name: certificationConsumerGroupName1
+    }
+    resource eventHubPubsubConsumerGroup2 'consumergroups' = {
+      name: certificationConsumerGroupName2
+    }
+  }
+  resource certificationEventHubPubsub2 'eventhubs' = {
```
Comment (berndverst) [inline .github/infrastructure/conformance/azure/conf-test-azure-eventhubs.bicep:109]: You never seem to use this. Given that, can you please remove this and rename `Pubsub1` to something more meaningful?

Especially where the variable is exported - can you name it `AzureEventHubsPubsubCertificationTopicConnectionString` or something along these lines? We already have enough topic used for conformance so it's confusing to have a random Pubsub1.. but there is no Pubsub2 being used.

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+   - App: Simulates periodic errors
+   - Component: Retries on error
+   - App: Observes successful messages
+   - Test: Confirms that all expected messages were received
+   - Test: Confirms that all expected system properties are set
+### Authentication Tests 
+  - Test connection string based authentication mechanism
+  - Test AAD Service Principal based authentication
+    - Utilize a service principal with appropriate roles granted
+  - Test MSI based authentication : [TODO]
```
Comment (berndverst) [inline tests/certification/pubsub/azure/eventhubs/README.md:69]: Ignore MSI -- you can remove this. I have decided that it isn't necessary -- given the libraries we use, if things work with service principals they will also work with MSI.

File: tests/certification/pubsub/azure/eventhubs/README.md [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    - Utilize Managed Identity
+### Network Tests?
```
Comment (berndverst) [inline tests/certification/pubsub/azure/eventhubs/README.md:71]: remove question mark

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (berndverst) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:290]: There will be concurrent runs of this workflow - how is that supposed to work?

**Review by berndverst — COMMENTED** [review]
LGTM so far, but I worry about what happens if this test is run multiple times in parallel - that's something to consider. Your cleanup step may break then.

For the IoT hub stuff - the azure CLI is available in the GitHub workflow and should have necessary credentials to publish to the IoT hub. You can enable the test and give it a try.

Please also take a look at the naming of variables.. I see PubSub1.. but PubSub2 is defined and unused. Let's use a better name then.

Once you make those last tweaks with variable naming as I mentioned I will apply this change against our infrastructure, from that point forward you can directly run your tests on this PR with `/ok-to-test`

**Review by berndverst — CHANGES_REQUESTED** [review]
Just so nobody merges this yet, I will request these changes :)

Let's add the IoT Hub tests in, and let's make sure everything passes with `/ok-to-test` before merging.

Also let's add the network interruption tests.

File: tests/certification/pubsub/azure/eventhubs/eventhubs_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/eventhubs_test.go:290]: In local debug, it worked fine. I was not aware that multiple flows can run in parallel. 

Good catch @berndverst will move this to initial flow itself as a last step in event management test.

File: .github/infrastructure/conformance/azure/conf-test-azure-eventhubs.bicep (hunk 16 above)
Comment (surenderssm) [reply .github/infrastructure/conformance/azure/conf-test-azure-eventhubs.bicep:109]: `certificationEventHubPubsub2` is being leveraged as part of following tests

- Publishing to multiple topics from the component with namespace level access.
- Publishing to a topic which is not being subscribed by any consumer.

Good point on naming convention, let me give it a shot.

File: tests/certification/pubsub/azure/eventhubs/README.md (hunk 18 above)
Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/README.md]: done.

File: tests/certification/pubsub/azure/eventhubs/delete-eventhub.sh [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------
+
+set -e
+
+# This script delete the Azure Eventhub created in the scoped Azure EventHub namespace as per the test suite.
+# Step 1 : Login to Azure with provided SPN which has `Azure Event Hubs Data Owner` role on scoped eventhub namespace.
+# Step 2 : Delete the eventhub/topic which is provided as an argument
```
Comment (surenderssm) [inline tests/certification/pubsub/azure/eventhubs/delete-eventhub.sh:18]: Deleting the checkpoint container related to this eventhub and consumer will add to the reliability of the test.

Comment (surenderssm) [reply tests/certification/pubsub/azure/eventhubs/delete-eventhub.sh:18]: taken care.

**Review by berndverst — CHANGES_REQUESTED** [review]
Test infrastructure changes have been applied (after I made some updated). I tried running the tests but some permission seems to be missing for the entity management case.

`AzureCertificationServicePrincipalClientId` probably requires an additional permission to create the topics (EventHubs).

From the embedded sidecar logs:
```
time="2022-04-24T01:06:25.161303172Z" level=info msg="app is subscribed to the following topics: [certification-topic-per-test-run] through pubsub=messagebus" instance=codespaces_4687ec scope=dapr.runtime type=log ver=unknown
time="2022-04-24T01:06:27.827773556Z" level=error msg="error creating event hub certification-topic-per-test-run: error code: 401, Details: Authorization failed for specified action: Manage,EntityWrite. TrackingId:eb25326b-c969-4340-8d38-6f78e98bb668_G18, SystemTracker:dapr3-conf-test-eventhubs.servicebus.windows.net:certification-topic-per-test-run, Timestamp:2022-04-24T01:06:27" instance=codespaces_4687ec scope=dapr.components type=log ver=unknown
time="2022-04-24T01:06:27.827821856Z" level=error msg="failed to subscribe to topic certification-topic-per-test-run: error creating event hub certification-topic-per-test-run: error code: 401, Details: Authorization failed for specified action: Manage,EntityWrite. TrackingId:eb25326b-c969-4340-8d38-6f78e98bb668_G18, SystemTracker:dapr3-conf-test-eventhubs.servicebus.windows.net:certification-topic-per-test-run, Timestamp:2022-04-24T01:06:27" instance=codespaces_4687ec scope=dapr.runtime type=log ver=unknown
```

**Comment by berndverst** [comment]
Let me know if you can figure out which permission is missing and I will apply that additional permission to our infrastructure -- please also update the shell script in this PR to grant the additional permission. @surenderssm

**Comment by surenderssm** [comment]
> Let me know if you can figure out which permission is missing and I will apply that additional permission to our infrastructure -- please also update the shell script in this PR to grant the additional permission. @surenderssm

This has been fixed, `Azure Event Hubs Data Owner` is the role which was required. This role was being assigned, however with an invalid scope path. Same has been fixed.

Scope path before the change 

`:/subscriptions/${SUB_ID}/resourceGroups/${RESOURCE_GROUP_NAME}/providers/Microsoft.EventHub/namespaces/${EVENT_HUBS_NAMESPACE}/eventhubs/*`

Scope path after the change :

`/subscriptions/${SUB_ID}/resourceGroups/${RESOURCE_GROUP_NAME}/providers/Microsoft.EventHub/namespaces/${EVENT_HUBS_NAMESPACE}`

**Review by berndverst — APPROVED** [review]

**Comment by berndverst** [comment]
I applied the new role to the infrastructure.

All tests pass now.
